internal/overlord/servstate/manager.go

```diff
@@ -39,34 +49,132 @@ func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string) (*Se
 	return manager, nil
 }
 
-func (m *ServiceManager) reloadSetup() error {
-	setup, err := setup.ReadDir(m.pebbleDir)
+func (m *ServiceManager) reloadPlan() error {
+	p, err := plan.ReadDir(m.pebbleDir)
+	if err != nil {
+		return err
+	}
+	m.plan = p
+	return nil
+}
+
+// Plan returns the configuration plan.
+func (m *ServiceManager) Plan() (*plan.Plan, error) {
+	releasePlan, err := m.acquirePlan()
+	if err != nil {
+		return nil, err
+	}
+	defer releasePlan()
+	return m.plan, nil
+}
+
+// AppendLayer appends the given layer to the plan's layers and updates the
+// layer.Order field to the new order. If a layer with layer.Label already
+// exists, return an error of type *LabelExists.
+func (m *ServiceManager) AppendLayer(layer *plan.Layer) error {
+	releasePlan, err := m.acquirePlan()
+	if err != nil {
+		return err
+	}
+	defer releasePlan()
+
+	index, _ := findLayer(m.plan.Layers, layer.Label)
+	if index >= 0 {
+		return &LabelExists{Label: layer.Label}
+	}
+
+	return m.appendLayer(layer)
+}
+
+func (m *ServiceManager) appendLayer(layer *plan.Layer) error {
+	newOrder := 1
+	if len(m.plan.Layers) > 0 {
+		last := m.plan.Layers[len(m.plan.Layers)-1]
+		newOrder = last.Order + 1
+	}
+
+	newLayers := append(m.plan.Layers, layer)
+	err := m.updateLayers(newLayers)
+	if err != nil {
+		return err
+	}
+	layer.Order = newOrder
+	return nil
+}
+
+func (m *ServiceManager) updateLayers(layers []*plan.Layer) error {
```

Comment:
I wonder if this might be a bit more clear if named updatePlan. On the first pass I was a bit confused when reading appendLayer calling updateLayers. Not a big deal, of course.

Reply:
Agreed, done.